### PR TITLE
c64 65816 extra banks emd

### PR DESCRIPTION
**Pull request by blackystardust**
This emd works on a c64 with a 65816 and extra banks, it has been confirmed to work on a scpu64, not tested on any other 65816 cards (since I don't have any of them).

### CODE REVIEW

File: libsrc/c64/emd/c64-65816.s [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        lda     (ptr4),y                ; get low byte of count
+        bne     @nodex
+        dex
+@nodex:
+.P816
+        dec
+.P02
+        sta     ptr3                    ; length low
+        txa
+        sta     ptr3+1                  ; length high
```
Comment (greg-king5) [inline libsrc/c64/emd/c64-65816.s]: Now that those `eor #$ff` instructions are gone, you can store the X register directly into `ptr3+1`.

**Comment by oliverschmidt** [comment]
These additional drivers (and the general approach behind putting them here) are welcome.

However if one of the community members takes the time to review the code and suggests an improvement the pull request author is supposed to take the time to react on that suggestion.

If we can't agree on that process then it doesn't make sense to have this pull request open any longer.

**Comment by mrdudz** [comment]
IMHO this one demonstrates nicely how this process is flawed.....

ie, it should have been merged long ago, and greg should have made this trivial improvement all by himself. its a novel idea to keep pull request open until they are "perfect" - but very impractical too.

(blacky has given up because of the longish process for that matter)

**Comment by oliverschmidt** [comment]
Hi,

> IMHO this one demonstrates nicely how this process is flawed.....
>
> ie, it should have been merged long ago, and greg should have made this
> trivial improvement all by himself. its a novel idea to keep pull request
> open until they are "perfect"
>
I certainly disagree. It's the process for all GitHub projects I know.

> - but very impractical too.
>
I understand your perspective but you don't understand that for many
projects the bottleneck is the maintainer, not the contributor. And from
that perspective it makes totally sense to move as much effort as possible
away from the bottleneck.

You may argue that in this specific case the effort for commenting was
larger than for making the actual change. But blacky planned to provide
(much) more code so it was reasonable to "invest into him" to have him
provide pull request ready to merge as-is.

You may additionally argue that it doesn't makes sense to be so strict
regarding minimal derivations from coding style rules etc when that leads
to valuable contributions not ending up in the repo at all. I'm willing to
own that this is just my personal style to handle things. However cc65 is
open for everybody to fork and become the upstream if the community
considers it to be that. In fact it's much more open than it ever was with
Uz...

Or maybe you want to join in and provide those minimal changes necessary to
merge a pull request as-is. However I personally will rather drop the
project altogether than become the scrubwoman for the code of contributors.

> (blacky has given up because of the longish process for that matter)
>
Don't get me wrong. The original idea with more drivers for cc65 and better
tests for VICE was great - in fact it still is. Thanks for your effort to
make the proposal! But I'm not sitting here waiting for cc65 contributors
to show up. I've (too many) other things to take care of so I spend time
for cc65 as I can and see fit. Everybody feeling that he's not only able
but also willing to actually make it better is welcome to try. As pointed
out above he doesn't even need me to invite him...

Regards,
Oliver

**Comment by mrdudz** [comment]
writing all this took longer than merging the change AND making the proposed fix would do.

this is the whole problem in a nutshell. really.

**Comment by oliverschmidt** [comment]
Hi,

> writing all this took longer than merging the change AND making the
> proposed fix would do.
>
What makes you believe that I'm not fully aware of that without you telling
me?

> this is the whole problem in a nutshell. really.
>
The whole problem is that people like you don't (want to) see the problem.
Threads like this make projects like cc65 an orphan (again) sooner or
later. That's the whole problem.

Regards,
Oliver

**Comment by mrdudz** [comment]
"people like me" want to contribute to the code and not waste time on discussing trivial changes. in fact i am pretty sure that none of use wants to waste time on discussing trivial changes. so what purpose does doing it anyway serve then? so far it drove away one very productive contributor, and kept a nice and welcomed addition from being merged. i am sure that's not what you are trying to achieve here.
seriously, "educating" contributors this way does not work. patches will hardly ever be "perfect". and frequently it will be the much better choice to just do some trivial fixes and apply the damn thing and be done with it, rather than wasting resources on a pointless discussion.
give a couple of regular contributors from the past the necessary rights to merge pull requests, and the whole problem wouldnt exist at all. i'd be happy to make a one line fix and approve requests like this if that keeps people like blacky from running away :)

**Comment by oliverschmidt** [comment]
Hi,

> "people like me" want to contribute to the code and not waste time on
> discussing trivial changes. in fact i am pretty sure that none of use wants
> to waste time on discussing trivial changes. so what purpose does doing it
> anyway serve then?
>
The point I was trying to make is that it's like a apartment-sharing
community where some live there and are just happy while the others are
busy cleaning up just because they happen want things (a little) clean than
the others. Saying "I don't see your problem, for me it's clean enough. If
you want it cleaner than that's your problem." doesn't solve the problem.

> so far it drove away one very productive contributor, and kept a nice and
> welcomed addition from being merged.
>
I absolutely follow you.

> i am sure that's not what you are trying to achieve here.
>
No but I need to balance between cc65 going without that contribution and
cc65 going without a maintainer. Have you checked Uz' SVN logs in how many
cases he reworked contributions? You see: These weren't contributions he
was waiting for. He had his own ideas what to do next. But instead he was
urged to clean up something someone put into his repo in order to still
"like what he saw" when looking at things.

I believe that this was at least one of the aspects that made he stop
running cc65 as before. We learned the hard way that he by no means wasn't
interested in cc65 anymore. He was just not interested in doing it with
"us" anymore. Think about that for a moment.

I believe that I do a better service to the cc65 community if I don't get
myself in the same position. Even if that means that individual
contributions may get lost - or unknown contributions are never implemented
because of the contribution process as it stands.

> seriously, "educating" contributors this way does not work. patches will
> hardly ever be "perfect".
>
I generally try to stay clear from appearing as the "know it all". In fact
it would be pretty stupid trying to do so as I know nearly nothing about
the cc65 core component - the compiler. However I won't dispute that it
easily looks that way to contributors.

> and frequently it will be the much better choice to just do some trivial
> fixes and apply the damn thing and be done with it, rather than wasting
> resources on a pointless discussion.
>
That's the point we both disagree. It's not about the time I invest, it's
about my very personal perspective on the time I invest.

> give a couple of regular contributors from the past the necessary rights
> to merge pull requests, and the whole problem wouldnt exist at all. i'd be
> happy to make a one line fix and approve requests like this if that keeps
> people like blacky from running away :)
>
Cool!

So far I don't remember anybody coming up and asking me to provide those
rights to him. In fact I offered it once and it was declined.

Above I was writing about the difference between my effort and my personal
perspective on effort. Maybe it's not that different here with you. It's
actually really not much effort to grab some 3rd party pull request, apply
a small change to it and provide an own pull request. But maybe that's
something you "just don't want" to do. You see: Everybody has things he
feels he just doesn't want to do - independently from the actual effort ;-)

So if for you the difference is that you can make fixes with direct pushes
instead of pull requests then here you go:

I just created a team "Developers" with write access to the cc65 source
repo (which includes the right to merge pull requests afaik) and made you
one of the members of that team.

I trust in your that you'll try to keep the content of the repo in a state
that I continue to "like what I see" when looking at things.

Welcome,
Oliver

**Comment by silverdr** [comment]
Hi fellows (Olliver excluded) - please just learn how this things work these days. Just run a few projects, where you are a maintainer/manager and see why this is done so. Or be a developer in such projects and see that you'll *always* have to make required changes to your "pull requests"/"merge requests" (as they are called in other tools) all by yourself. Well run projects don't let any of your code into the master branch without thorough review. Reviewers spend their time pointing out things that need to be corrected. They DON'T spend time correcting your code. That's how it works. I am sorry to hear that a potentially valuable contributor left because he didn't understand the process. But that's how things work and are supposed to work these days. Get over it, learn it and you'll see by yourself that it's better than the "classic" ways.

**Comment by mrdudz** [comment]
> Just run a few projects, where you are a maintainer/manager

you must be kidding ^_^

**Comment by silverdr** [comment]
Since we all agree that the submission made by blackystardust is valuable and I can't update his pull request, I made both the original and the requested post-review changes to my own fork and submitted a new pull request. This way the valuable contribution is not lost. Took less than a quarter, including re-syncing my fork.

**Comment by mrdudz** [comment]
thanks, that saves me from wrestling with github this time :)

and since oliver granted me the right to do so i will now:
- close this PR (its already merged and thus pointless) and
- also merge this PR: https://github.com/cc65/cc65/pull/241 (all complaints fix and no new comments since almost 3 months. why it wasnt already merged seems weird to me =P)

**Comment by greg-king5** [comment]
My fault.  I was supposed to let Oliver know when I no longer had anything to say about both of those PRs.  But, I was "juggling alot of balls".  So, I neglected those two.

**Pull request closed without merging**